src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -382,14 +382,14 @@ internal int UnregisterGrainForTesting(GrainId grain)
 
 #region Grains
 
-        internal bool IsReentrantGrain(ActivationId running)
+        internal bool IsMessageInterleaves(Message message, ActivationId running)
         {
             ActivationData target;
             GrainTypeData data;
             return TryGetActivationData(running, out target) &&
                 target.GrainInstance != null &&
                 GrainTypeManager.TryGetData(TypeUtils.GetFullName(target.GrainInstanceType), out data) &&
-                data.IsReentrant;
+                data.IsMessageInterleaves(message);
```

Comment:
I would leave data.IsReentrant and add OR data.IsMessageAllowedToInterleave.
That way you separate the 2 things, which may be useful in other places.
Plus naming: `IsMessageAllowedToInterleave`.